eZ/Publish/Core/REST/Server/Input/Parser/Criterion/LogicalAnd.php

```diff
@@ -30,15 +35,65 @@ class LogicalAnd extends CriterionParser
      */
     public function parse(array $data, ParsingDispatcher $parsingDispatcher)
     {
-        if (!array_key_exists('AND', $data) && !is_array($data['AND'])) {
-            throw new Exceptions\Parser('Invalid <AND> format');
+        if (!array_key_exists(static::TAG_NAME, $data) || !is_array($data[static::TAG_NAME])) {
+            throw new Exceptions\Parser('Invalid <' . static::TAG_NAME . '> format');
         }
 
         $criteria = array();
-        foreach ($data['AND'] as $criterionName => $criterionData) {
-            $criteria[] = $this->dispatchCriterion($criterionName, $criterionData, $parsingDispatcher);
+
+        $flattenedCriteriaElements = $this->getFlattenedCriteriaData($data[static::TAG_NAME]);
+        foreach ($flattenedCriteriaElements as $criterionElement) {
+            $criteria[] = $this->dispatchCriterion(
+                $criterionElement['type'],
+                $criterionElement['data'],
+                $parsingDispatcher
+            );
+        }
+
+        return new Values\Content\Query\Criterion\LogicalAnd($criteria);
+    }
+
+    /**
+     * @param array $criteriaByType
+     * @return array
+     */
+    protected function getFlattenedCriteriaData(array $criteriaByType)
+    {
+        $criteria = [];
+        foreach ($criteriaByType as $type => $criterion) {
+            if (is_array($criterion) && $this->isNumericArray($criterion)) {
+                foreach ($criterion as $criterionElement) {
+                    $criteria[] = [
+                        'type' => $type,
+                        'data' => $criterionElement,
+                    ];
+                }
+            } else {
+                $criteria[] = [
+                    'type' => $type,
+                    'data' => $criterion,
+                ];
+            }
+        }
+
+        return $criteria;
+    }
+
+    /**
+     * Checks if the given $value is a purely numeric array.
+     *
+     * @param array $value
+     *
+     * @return bool
+     */
+    protected function isNumericArray(array $value)
```

Comment:
Naming: `isZeroBasedArray` - this is what we actually should check. Any other type of data here should be handled as an associative array of data.

Reply:
Actually I implemented this in a different way in the first place but found this in 

`\eZ\Publish\Core\REST\Common\Output\Generator\Json\FieldTypeHashGenerator`

in line 121.

Reply:
To be honest I was confused about zeroBasedArray. For me a zero based array is an array starting at the index 0.

But this is not what I intended to check here. I want to check if all elements have numeric indexes.

What I am actually checking is a numeric/sequential array with purely numeric keys for all elements, the opposite would be an associative array in my opinion.

`$this->isNumericArray(['a','b','c']); // returns true`
`$this->isNumericArray(['a' => 'a', 'b' => 'b', 'c' => 'c']); // returns false`
`$this->isNumericArray([0 => 'a', 'some_string' => 'a', 'some_other_string' => 'b'); // returns false`

So in order to be 100% sure I ask one more time if this is clear. Hope I'm not annoying you with that.

`Any other type of data here should be handled as an associative array of data.`
In your suggestion of the next comment you treat
```
[
  0 => '1234',
  'some_string' => '1234'
]
```
as a non associative array, although..
I think that this is anyway not possible to happen on the xml processing of EZ and we can be sure that the array are either full numeric OR associative.

Reply:
> To be honest I was confused about zeroBasedArray. For me a zero based array is an array starting at the index 0.
> But this is not what I intended to check here. I want to check if all elements have numeric indexes.

The name "zero-based" for array implies the other keys are numerical as well. I get that what I've requested does not necessarily check that, but in this case there's no other input possible. Even if someone decides to call his Criterion `'0'`, it's still gonna be a `string`. So it's safe to assume that if the first key of an array is integer zero, the whole array contains numeric keys. Unless I missed something when looking over Input Parser code :)

> I think that this is anyway not possible to happen on the xml processing of EZ and we can be sure that the array are either full numeric OR associative.

My point exactly :)